kad/src/high_level.rs

```diff
@@ -18,513 +18,446 @@
+                    let multiaddrs = mem::replace(&mut peer.multiaddrs, Vec::new());
+                    trace!("Reporting multiaddresses for {:?}: {:?}", peer.node_id, multiaddrs);
+                    new_known_multiaddrs.push((peer.node_id.clone(), multiaddrs));
+                }
+
+                if peer.node_id.distance_with(&searched_key)
+                    <= state.result[0].distance_with(&searched_key)
+                {
+                    local_nearest_node_updated = true;
+                }
+
+                if state.result.iter().any(|ma| ma == &peer.node_id) {
+                    continue;
+                }
+
+                // Insert the node into `pending_nodes` at the right position, or do not
+                // insert it if it is already in there.
+                if let Some(insert_pos) = state.pending_nodes.iter().position(|e| {
+                    e.distance_with(&searched_key) >= peer.node_id.distance_with(&searched_key)
+                }) {
+                    if state.pending_nodes[insert_pos] != peer.node_id {
+                        state.pending_nodes.insert(insert_pos, peer.node_id.clone());
                     }
+                } else {
+                    state.pending_nodes.push(peer.node_id.clone());
                 }
-                match entry.insert(Connection::Pending(Vec::with_capacity(1))) {
-                    &mut Connection::Pending(ref mut list) => list,
-                    _ => unreachable!("we just inserted a Pending variant"),
+            }
+
+            if state.result.len() >= num_results
+                || (state.stage != Stage::FirstStep && state.current_attempts_fut.is_empty())
+            {
+                state.stage = Stage::FinishingNextIter;
+
+            } else {
+                if !local_nearest_node_updated {
+                    trace!("Loop didn't update closer node ; jumping to step 2");
+                    state.stage = Stage::SecondStep;
                 }
             }
-        };
 
-        let (tx, rx) = oneshot::channel();
-        let mut tx = Some(tx);
-        let mut and_then = Some(and_then);
-        pending_list.push(Box::new(move |ctrl: &mut KademliaServerController| {
-            let and_then = and_then
-                .take()
-                .expect("Programmer error: 'pending' closure was called multiple times");
-            let tx = tx.take()
-                .expect("Programmer error: 'pending' closure was called multiple times");
-            let ret = and_then(ctrl);
-            let _ = tx.send(ret);
-        }) as Box<_>);
-
-        let future = rx.map_err(|_| IoErrorKind::ConnectionAborted.into());
-        let future = self.inner.timer.timeout(future, self.inner.timeout);
-        Box::new(future) as Box<_>
-    }
+            future::ok((Some(KadQueryEvent::NewKnownMultiaddrs(new_known_multiaddrs)), state))
+        });
+
+        Some(future::Either::B(future))
+    }).filter_map(|val| val);
+
+    Box::new(stream) as Box<_>
```

Comment:
`stream` does not need to be boxed.

Reply:
I decided to `Box` it anyway in order to improve the compilation time.
I've added a comment about this.
But if you think it shouldn't be boxed, I can remove the boxing.